src/shared/clients/s3Client.ts

```diff
@@ -0,0 +1,94 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+
+export const DEFAULT_MAX_KEYS = 300
+export const DEFAULT_DELIMITER = '/'
+
+export interface S3Client {
+    /**
+     * Creates a bucket in the region of the client.
+     */
+    createBucket(request: CreateBucketRequest): Promise<CreateBucketResponse>
+
+    listBuckets(): Promise<ListBucketsResponse>
+
+    listObjects(request: ListObjectsRequest): Promise<ListObjectsResponse>
+
+    createFolder(request: CreateFolderRequest): Promise<CreateFolderResponse>
+
+    downloadFile(request: DownloadFileRequest): Promise<void>
+    uploadFile(request: UploadFileRequest): Promise<void>
+}
+
+export interface Bucket {
+    readonly name: string
+    readonly region: string
+    readonly arn: string
+}
```

Comment:
Does the SDK not already provide such interfaces? Why do we need to wrap them?

Reply:
Not really. It just does request / response. S3 also provides a lot more information we don't use and don't provide `region` and `arn`.

```ts
  export interface CreateBucketRequest {
    ACL?: BucketCannedACL;
    Bucket: BucketName;
    CreateBucketConfiguration?: CreateBucketConfiguration;
    GrantFullControl?: GrantFullControl;
    GrantRead?: GrantRead;
    GrantReadACP?: GrantReadACP;
    GrantWrite?: GrantWrite;
    GrantWriteACP?: GrantWriteACP;
    ObjectLockEnabledForBucket?: ObjectLockEnabledForBucket;
  }
```

My approach to SDK wrapping is basically [(stuff I think the SDK should have) - (stuff we don't need)] customized to our domain and made testable.

Reply:
What's wrong with having extra info? 

Is this worth  the extra code (interfaces + tests + mapping routines + ...). Also we get less re-use because the types are incompatible... and the names are different for the same concept.

Reply:
There are quite a few pros to slimming and customizing the interfaces:

* Tests are easier to write since you don't need to pass around stuff you don't use
* The interface contract is clearer for your callers since you specify exactly what you depend on or return
* You actually get more re-usability because other parts of the code that do similar things would end up re-implementing the same wrapper logic to adapt to the domain, make it testable, etc.
* You can add functionality specific to your domain or rename/remove functionality that doesn't apply

It's basically the [interface segregation principle](https://en.wikipedia.org/wiki/Interface_segregation_principle). The wrapper takes just what it needs and shields its consumers from any data/complexity they don't need to know about.

Of course there are cons as well, but it certainly makes things cleaner.

Reply:
> Tests are easier to write since you don't need to pass around stuff you don't use

Why do we need to test fields we don't use that are already tested by the SDK?

> The interface contract is clearer for your callers since you specify exactly what you depend on or return

The callers don't really care?

> You actually get more re-usability because other parts of the code that do similar things would end up re-implementing the same wrapper logic to adapt to the domain, make it testable, etc.

If we never wrap we never reuse?

> You can add functionality specific to your domain or rename/remove functionality that doesn't apply

Why do we need to remove anything? functions are OK?

Reply:
> Why do we need to test fields we don't use that are already tested by the SDK?

It's not actually the matter of testing the fields. It's the matter of having to populate (and ignore during verification) the fields you don't use when you write tests simply because they exist (and may be required / validated).

i.e. if you're writing a test that for the caller that sets `foo` in the request and reads `bar` in the response. All you really want to know is whether you set foo properly and whether you received bar properly.

if you're writing a fake for the request and response, all you really want to fake is `foo` and `bar.

> The callers don't really care?

The smaller interface is easier to integrate with and write tests against. If you've had to write fakes / mocks for a large interface that you barely make use of, you can see the benefit. At some point you need to narrow down to the parts of the interface that you actually use. Rather than every caller needing to do that, it's nicer to abstract this at the earliest point in the code.

> If we never wrap we never reuse?

This is speaking to the reuse of the abstraction. The narrowing of the interface mentioned above.

> Why do we need to remove anything? functions are OK?

To lower the cognitive load on the caller and lower the surface area that you need to test.

Reply:
I really like the concept of a `Bucket` as it's own entity - but this implementation is literally just a data class - which seems pointless to me. I'd be in favour if `Bucket` was something you could actually *do* something with because it encapsulated the underlying S3 client and bucket name. The JB equivalent is [`S3VirtualBucket`](https://github.com/aws/aws-toolkit-jetbrains/blob/master/jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3VirtualBucket.kt); once you've got a reference to one of those - you can do common bucket operations like listing objects (`S3VirtualBucket#listObjects()`), delete, rename, upload etc without ever having to care about the region, client, bucket name etc - cos that's all encapsulated in the Bucket object itself.

The above `Bucket` doesn't really help us do anything - it's literally just a data class - and like others have already mentioned that feels like a pointless maintenance treadmill without much benefit.

Reply:
Discussed offline. The `Bucket` is actually a container missing from the SDK. Other containers are either missing or added here just for consistency rather than mixing SDK and custom types.

A "smart" object that abstracts operations is certainly a valid pattern similar to a [DAO](https://en.wikipedia.org/wiki/Data_access_object). I think a stronger case could be made for it if we were using these throughout the code base. (e.g. we would have `Bucket.list()`,  `LambdaFunction.getByArn('...')`, `lambdaFunction.update('..')`, etc.). Also if the AWS SDK were using this pattern, it would be a really good reason to adopt it. I have indeed seen some SDKs that do this.

However, we're just passing around (thin wrapper) clients that return "dumb" plain old data objects. I actually prefer this approach for the testability benefits. Not as elegant though. A case could certainly be made for either.